substate-store-impls/src/hash_tree/mod.rs

```diff
@@ -38,372 +37,49 @@ mod types;
-            },
-        }
-    }
-
-    pub fn for_substate_reset(sort_key: &DbSortKey, value: &'a DbSubstateValue) -> Self {
-        Self {
-            key_bytes: sort_key.0.clone(),
-            new_payload: Some(SubstateValue(value)),
-        }
-    }
-}
-
-fn put_leaf_hash_changes<S: TreeStore>(
-    store: &S,
-    current_version: Option<Version>,
-    next_version: Version,
-    changes: Vec<LeafChange>,
-) -> Option<Hash> {
-    // We need to re-allocate it like this in order to:
-    // - construct the right shape (i.e. a reference to a tuple) of items passed to JMT,
-    // - and recover Substate values based on `LeafKey` from the update batch returned by JMT.
-    let changes = changes
-        .into_iter()
-        .map(|change| {
-            (
-                LeafKey {
-                    bytes: change.key_bytes,
-                },
-                change
-                    .new_payload
-                    .map(|payload| ((payload.value_hash(), next_version), payload)),
-            )
-        })
-        .collect::<IndexMap<_, _>>();
-    let (root_hash, update_result) = JellyfishMerkleTree::new(store)
-        .batch_put_value_set(
-            changes
-                .iter()
-                .map(|(key, change)| {
-                    (
-                        key,
-                        change
-                            .as_ref()
-                            .map(|(hash_and_version, _payload)| hash_and_version),
-                    )
-                })
-                .collect(),
-            None,
-            current_version,
-            next_version,
-        )
-        .expect("error while reading tree during put");
-
-    for (key, node) in update_result.node_batch.into_iter().flatten() {
-        // We promised to associate Substate values; but not all newly-created nodes are leaves:
-        if let Node::Leaf(leaf_node) = &node {
-            // And not every newly-created leaf comes from a value change: (sometimes it is just a tree re-structuring!)
-            if let Some(change) = changes.get(leaf_node.leaf_key()) {
-                // Now: if a JMT leaf was created due to value change, then it must have been an upsert:
-                let (_hash_and_version, payload) = change.as_ref().expect("unexpected delete");
-                // Only the Substate Tier's leaves contain Substate values:
-                if let Payload::SubstateValue(substate_value) = payload {
-                    store.associate_substate_value(&key, *substate_value);
-                }
-            }
-        }
-        let node = TreeNode::from(&key, &node);
-        store.insert_node(key, node);
-    }
-    for key in update_result.stale_node_index_batch.into_iter().flatten() {
-        store.record_stale_tree_part(StaleTreePart::Node(key.node_key));
-    }
-
-    if root_hash == SPARSE_MERKLE_PLACEHOLDER_HASH {
-        None
-    } else {
-        Some(root_hash)
-    }
-}
-
-pub struct NestedTreeStore<S> {
```

Comment:
I see you "implement" this responsibility kinda integrated with the "tier framework". I can live with it, because it is ~3 lines.

However, I used this struct also in the Node (see https://github.com/radixdlt/babylon-node/pull/851), to do some custom tree-walking. And it allowed a "Really clean impl ❤️" (according to the reviewer), so probably worth keeping?

Would you be able to save the `NestedTreeStore` from oblivion?
I see that 2 lower `Tier` structs have fields `base_store + tree_node_prefix` right next to each other, which secretly is a 100% of `NestedTreeStore` - wouldn't it be nicer to explicitly use a `NestedTreeStore` impl detail instead?

Reply:
As discussed on the call, I think it's better if we lean into the typed tiers for the substate key iteration :)

Reply:
That said, I think I agree with you that there's actually a "TypedNestedTree<K: TypedLeafKey, P: TypedLeafPayload>" hidden inside each of these tiers which could be extracted / composed (and basically implements what currently the `Tier` trait is) - this would operate at a lower abstraction level.

Which would then just leave the business logic in the `EntityTier` / `PartitionTier` / `SubstateTier`.

This would be another small/medium refactoring on top of what's currently here.